File: phases/10-llms-from-scratch/01-tokenizers/code/bpe.py
from collections import Counter


class BPETokenizer:
    def __init__(self):
        self.merges = {}
        self.vocab = {}

    def _get_pairs(self, tokens):
        pairs = Counter()
        for i in range(len(tokens) - 1):
            pairs[(tokens[i], tokens[i + 1])] += 1
        return pairs

    def _merge_pair(self, tokens, pair, new_token):
        merged = []
        i = 0
        while i < len(tokens):
            if i < len(tokens) - 1 and tokens[i] == pair[0] and tokens[i + 1] == pair[1]:
                merged.append(new_token)
                i += 2
            else:
                merged.append(tokens[i])
                i += 1
        return merged

    def train(self, text, num_merges):
        tokens = list(text.encode("utf-8"))
        self.vocab = {i: bytes([i]) for i in range(256)}

        for i in range(num_merges):
            pairs = self._get_pairs(tokens)
            if not pairs:
                break
            best_pair = max(pairs, key=pairs.get)
            new_token = 256 + i
            tokens = self._merge_pair(tokens, best_pair, new_token)
            self.merges[best_pair] = new_token
            self.vocab[new_token] = self.vocab[best_pair[0]] + self.vocab[best_pair[1]]
            merged_str = self.vocab[new_token]
            print(f"Merge {i + 1}: {best_pair} -> {new_token} = {merged_str}")

        return self

    def encode(self, text):
        tokens = list(text.encode("utf-8"))
        for pair, new_token in self.merges.items():
            tokens = self._merge_pair(tokens, pair, new_token)
        return tokens

    def decode(self, tokens):
        byte_sequence = b"".join(self.vocab[t] for t in tokens)
        return byte_sequence.decode("utf-8", errors="replace")

    def vocab_size(self):
        return len(self.vocab)

    def get_token_str(self, token_id):
        return self.vocab.get(token_id, b"<?>")


def demo_bpe():
    corpus = (
        "The cat sat on the mat. The cat ate the rat. "
        "The dog sat on the log. The dog ate the frog. "
        "Natural language processing is the study of how computers "
        "understand and generate human language."
    )

    print("=" * 60)
    print("Training BPE tokenizer")
    print("=" * 60)

    tokenizer = BPETokenizer()
    tokenizer.train(corpus, num_merges=30)

    print(f"\nVocabulary size: {tokenizer.vocab_size()}")

    test_sentences = [
        "The cat sat on the mat.",
        "The frog sat on the log.",
        "language processing",
        "unhappiness",
    ]

    print("\n" + "=" * 60)
    print("Encoding test sentences")
    print("=" * 60)

    for sentence in test_sentences:
        encoded = tokenizer.encode(sentence)
        decoded = tokenizer.decode(encoded)
        raw_bytes = len(sentence.encode("utf-8"))
        print(f"\nOriginal:  {sentence}")
        print(f"Encoded:   {encoded}")
        print(f"Decoded:   {decoded}")
        print(f"Tokens:    {len(encoded)} (from {raw_bytes} bytes)")
        print(f"Ratio:     {len(encoded) / raw_bytes:.2f}")


def demo_tiktoken():
    try:
        import tiktoken
    except ImportError:
        print("\ntiktoken not installed. Run: pip install tiktoken")
        return

    print("\n" + "=" * 60)
    print("Comparing with tiktoken (GPT-4 tokenizer)")
    print("=" * 60)

    enc = tiktoken.get_encoding("cl100k_base")

    test_texts = [
        "The cat sat on the mat.",
        "unhappiness",
        "Hello, world!",
        "def fibonacci(n):",
        "3.14159265358979",
    ]

    for text in test_texts:
        tokens = enc.encode(text)
        decoded_pieces = [enc.decode([t]) for t in tokens]
        print(f"\n'{text}'")
        print(f"  Tokens:  {decoded_pieces}")
        print(f"  IDs:     {tokens}")
        print(f"  Count:   {len(tokens)}")


if __name__ == "__main__":
    demo_bpe()
    demo_tiktoken()

File: phases/10-llms-from-scratch/01-tokenizers/code/bpe.rs
use std::collections::HashMap;

struct BPETokenizer {
    merges: Vec<((u32, u32), u32)>,
    vocab: HashMap<u32, Vec<u8>>,
}

impl BPETokenizer {
    fn new() -> Self {
        let mut vocab = HashMap::new();
        for i in 0u32..256 {
            vocab.insert(i, vec![i as u8]);
        }
        BPETokenizer {
            merges: Vec::new(),
            vocab,
        }
    }

    fn get_pairs(tokens: &[u32]) -> HashMap<(u32, u32), usize> {
        let mut pairs = HashMap::new();
        for window in tokens.windows(2) {
            *pairs.entry((window[0], window[1])).or_insert(0) += 1;
        }
        pairs
    }

    fn merge_pair(tokens: &[u32], pair: (u32, u32), new_token: u32) -> Vec<u32> {
        let mut merged = Vec::with_capacity(tokens.len());
        let mut i = 0;
        while i < tokens.len() {
            if i < tokens.len() - 1 && tokens[i] == pair.0 && tokens[i + 1] == pair.1 {
                merged.push(new_token);
                i += 2;
            } else {
                merged.push(tokens[i]);
                i += 1;
            }
        }
        merged
    }

    fn train(&mut self, text: &str, num_merges: usize) {
        let mut tokens: Vec<u32> = text.as_bytes().iter().map(|&b| b as u32).collect();

        for i in 0..num_merges {
            let pairs = Self::get_pairs(&tokens);
            if pairs.is_empty() {
                break;
            }

            let best_pair = *pairs.iter().max_by_key(|&(_, count)| count).unwrap().0;
            let new_token = 256 + i as u32;

            tokens = Self::merge_pair(&tokens, best_pair, new_token);

            let mut new_bytes = self.vocab[&best_pair.0].clone();
            new_bytes.extend_from_slice(&self.vocab[&best_pair.1]);

            let display = String::from_utf8_lossy(&new_bytes);
            println!(
                "Merge {}: ({}, {}) -> {} = {:?}",
                i + 1,
                best_pair.0,
                best_pair.1,
                new_token,
                display
            );

            self.vocab.insert(new_token, new_bytes);
            self.merges.push((best_pair, new_token));
        }
    }

    fn encode(&self, text: &str) -> Vec<u32> {
        let mut tokens: Vec<u32> = text.as_bytes().iter().map(|&b| b as u32).collect();
        for &(pair, new_token) in &self.merges {
            tokens = Self::merge_pair(&tokens, pair, new_token);
        }
        tokens
    }

    fn decode(&self, tokens: &[u32]) -> String {
        let bytes: Vec<u8> = tokens
            .iter()
            .flat_map(|&t| self.vocab.get(&t).cloned().unwrap_or_else(|| vec![b'?']))
            .collect();
        String::from_utf8_lossy(&bytes).into_owned()
    }

    fn vocab_size(&self) -> usize {
        self.vocab.len()
    }
}

fn main() {
    let corpus = concat!(
        "The cat sat on the mat. The cat ate the rat. ",
        "The dog sat on the log. The dog ate the frog. ",
        "Natural language processing is the study of how computers ",
        "understand and generate human language."
    );

    println!("{}", "=".repeat(60));
    println!("Training BPE tokenizer");
    println!("{}", "=".repeat(60));

    let mut tokenizer = BPETokenizer::new();
    tokenizer.train(corpus, 30);

    println!("\nVocabulary size: {}", tokenizer.vocab_size());

    let test_sentences = vec![
        "The cat sat on the mat.",
        "The frog sat on the log.",
        "language processing",
        "unhappiness",
    ];

    println!("\n{}", "=".repeat(60));
    println!("Encoding test sentences");
    println!("{}", "=".repeat(60));

    for sentence in test_sentences {
        let encoded = tokenizer.encode(sentence);
        let decoded = tokenizer.decode(&encoded);
        let raw_bytes = sentence.len();

        println!("\nOriginal:  {}", sentence);
        println!("Encoded:   {:?}", encoded);
        println!("Decoded:   {}", decoded);
        println!("Tokens:    {} (from {} bytes)", encoded.len(), raw_bytes);
        println!("Ratio:     {:.2}", encoded.len() as f64 / raw_bytes as f64);
    }

    println!("\n{}", "=".repeat(60));
    println!("Performance: encode 100K iterations");
    println!("{}", "=".repeat(60));

    let test = "The cat sat on the mat and the dog sat on the log.";
    let start = std::time::Instant::now();
    for _ in 0..100_000 {
        let _ = tokenizer.encode(test);
    }
    let elapsed = start.elapsed();
    println!(
        "100K encodes in {:.2}ms ({:.0} encodes/sec)",
        elapsed.as_secs_f64() * 1000.0,
        100_000.0 / elapsed.as_secs_f64()
    );
}
